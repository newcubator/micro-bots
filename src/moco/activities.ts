import axios from "axios";
import { autoPage } from "./auto-page";
import { MOCO_TOKEN } from "./token";
import { MocoActivity } from "./types/moco-types";

/*
 * @See https://github.com/hundertzehn/mocoapp-api-docs/blob/master/sections/activities.md
 */

export async function getActivities(from: string, to: string) {
  return autoPage<MocoActivity>((page: number) => getActivitiesPaged(from, to, page));
}

const getActivitiesPaged = (from, to, page) =>
  axios.get<MocoActivity>("https://newcubator.mocoapp.com/api/v1/activities", {
    headers: {
      Authorization: "Token token=" + MOCO_TOKEN,
    },
    params: { from, to, page },
  });

export function getUserActivities(from: string, to: string, user_id: string) {
  return axios.get<MocoActivity[]>("https://newcubator.mocoapp.com/api/v1/activities", {
    headers: {
      Authorization: "Token token=" + MOCO_TOKEN,
    },
    params: {
      from,
      to,
      user_id,
    },
  });
}
