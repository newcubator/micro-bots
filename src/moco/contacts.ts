import axios from "axios";
import { MOCO_TOKEN } from "./token";
import { MocoContact } from "./types/moco-types";

export async function getContactById(id: string) {
  return axios
    .get<MocoContact>(`https://newcubator.mocoapp.com/api/v1/contacts/people/${id}`, {
      headers: {
        Authorization: "Token token=" + MOCO_TOKEN,
      },
    })
    .then((response) => response.data);
}
