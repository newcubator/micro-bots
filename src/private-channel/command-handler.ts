import { ActionType, PrivateChannelFields } from "../slack/types/slack-types";

export const commandHandler = async () => {
  return {
    statusCode: 200,
    body: JSON.stringify({
      text: "Private channel angefragt",
      blocks: [
        {
          type: "section",
          block_id: PrivateChannelFields.PRIVATE_CHANNEL_USERS,
          text: {
            type: "mrkdwn",
            text: "Ich erstelle dir gerne einen privaten Channel. Gib hierfür nur an, wen du nicht in dem Channel haben möchtest:",
          },
          accessory: {
            type: "multi_users_select",
            action_id: PrivateChannelFields.PRIVATE_CHANNEL_USERS,
            placeholder: {
              type: "plain_text",
              text: "Nutzer auswählen...",
            },
          },
        },
        {
          type: "input",
          block_id: PrivateChannelFields.PRIVATE_CHANNEL_NAME,
          element: {
            type: "plain_text_input",
            action_id: PrivateChannelFields.PRIVATE_CHANNEL_NAME,
            multiline: false,
            max_length: 200,
          },
          label: {
            type: "plain_text",
            text: "Channel Name (keine Leerzeichen und nur Kleinbuchstaben):",
            emoji: true,
          },
        },
        {
          type: "actions",
          block_id: "confirmationButton",
          elements: [
            {
              type: "button",
              text: {
                type: "plain_text",
                text: "Erstellen",
              },
              value: "Confirmation",
              style: "primary",
              action_id: ActionType.PRIVATE_CHANNEL,
            },
          ],
        },
      ],
    }),
  };
};
