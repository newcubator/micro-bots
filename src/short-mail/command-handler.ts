import { APIGatewayEvent } from "aws-lambda";
import { ActionType, ShortMailFields } from "../slack/types/slack-types";

export const commandHandler = async (event: APIGatewayEvent) => {
  return {
    statusCode: 200,
    body: JSON.stringify({
      response_type: "in_channel",
      text: "Kurzbrief angefragt",
      blocks: [
        {
          type: "input",
          block_id: ShortMailFields.SHORT_MAIL_RECIPIENT,
          label: {
            type: "plain_text",
            text: "Ich erstelle dir gerne einen Kurzbrief. Wähle dazu den Empfänger und deinen Standort aus und gibt deine Nachricht ein.",
          },
          element: {
            type: "external_select",
            action_id: ShortMailFields.SHORT_MAIL_RECIPIENT,
            placeholder: {
              type: "plain_text",
              text: "Empfänger auswählen...",
              emoji: true,
            },
          },
        },
        {
          type: "input",
          block_id: ShortMailFields.SHORT_MAIL_TEXT,
          element: {
            type: "plain_text_input",
            action_id: ShortMailFields.SHORT_MAIL_TEXT,
            multiline: true,
            max_length: 2500,
          },
          label: {
            type: "plain_text",
            text: "Deine Nachricht:",
            emoji: true,
          },
        },
        {
          type: "input",
          block_id: ShortMailFields.SHORT_MAIL_LOCATION,
          label: {
            type: "plain_text",
            text: "Wähle deinen Standort aus:",
          },
          element: {
            type: "static_select",
            action_id: ShortMailFields.SHORT_MAIL_LOCATION,
            placeholder: {
              type: "plain_text",
              text: "Standort auswählen...",
              emoji: true,
            },
            options: [
              {
                text: {
                  type: "plain_text",
                  text: "Dortmund",
                  emoji: true,
                },
                value: "D",
              },
              {
                text: {
                  type: "plain_text",
                  text: "Hannover",
                  emoji: true,
                },
                value: "H",
              },
            ],
          },
        },
        {
          type: "actions",
          block_id: "confirmationButton",
          elements: [
            {
              type: "button",
              text: {
                type: "plain_text",
                text: "Generieren",
              },
              value: "Confirmation",
              style: "primary",
              action_id: ActionType.SHORT_MAIL,
            },
          ],
        },
      ],
    }),
  };
};
