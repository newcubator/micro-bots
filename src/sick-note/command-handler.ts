import { APIGatewayEvent } from "aws-lambda";

export const commandHandler = async (event: APIGatewayEvent) => {
  return {
    statusCode: 200,
    body: JSON.stringify({
      text: "Krankschreibung angefragt",
      blocks: [
        {
          type: "input",
          block_id: "radio_buttons_days",
          element: {
            type: "radio_buttons",
            options: [
              {
                text: {
                  type: "plain_text",
                  text: "Ich möchte mich für heute krank melden",
                  emoji: true,
                },
                value: "single-day",
              },
              {
                text: {
                  type: "plain_text",
                  text: "Ich wurde krank geschrieben",
                  emoji: true,
                },
                value: "multiple-days",
              },
            ],
            action_id: "radio_buttons_action",
          },
          label: {
            type: "plain_text",
            text: "Krankmeldung",
            emoji: true,
          },
        },
        {
          type: "divider",
        },
        {
          type: "context",
          elements: [
            {
              type: "plain_text",
              text: "Von wann bis wann gilt deine AU?",
              emoji: true,
            },
          ],
        },
        {
          type: "actions",
          block_id: "dates",
          elements: [
            {
              type: "datepicker",
              placeholder: {
                type: "plain_text",
                text: "Start-Datum",
                emoji: true,
              },
              action_id: "start_date",
            },
            {
              type: "datepicker",
              placeholder: {
                type: "plain_text",
                text: "End-Datum",
                emoji: true,
              },
              action_id: "end_date",
            },
          ],
        },
        {
          type: "actions",
          elements: [
            {
              type: "button",
              text: {
                type: "plain_text",
                text: "Krank melden",
                emoji: true,
              },
              value: "Confirmation",
              style: "primary",
              action_id: "SICK_NOTE",
            },
          ],
        },
      ],
    }),
  };
};
