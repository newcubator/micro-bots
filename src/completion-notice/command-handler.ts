import { APIGatewayEvent } from "aws-lambda";
import { getProjects } from "../moco/projects";

export const commandHandler = async (event: APIGatewayEvent) => {
  // no need to parse the command input

  const projects = await getProjects({ include_archived: false });
  const options = projects
    .filter((project) => project.deal)
    .slice(0, 100) // slack allows 100 options max
    .map((project) => {
      return {
        value: project.id.toString(),
        text: {
          type: "plain_text",
          text: project.name,
          emoji: true,
        },
      };
    });
  console.log(`Loaded ${options.length} projects to select from`);

  if (!options.length) {
    return {
      statusCode: 200,
      body: JSON.stringify({
        response_type: "in_channel",
        text: "Konnte keine projekte finden.",
      }),
    };
  }

  return {
    statusCode: 200,
    body: JSON.stringify({
      response_type: "in_channel",
      blocks: [
        {
          type: "section",
          text: {
            type: "mrkdwn",
            text: "Ich erstelle dir gerne einen Stundenzettel. Wähle dazu bitte links das gewünschte Projekt aus und ich mache mich sofort an die Arbeit.",
          },
          accessory: {
            type: "static_select",
            action_id: "completionNotice",
            placeholder: {
              type: "plain_text",
              text: "Projekt auswählen...",
              emoji: true,
            },
            options: options,
          },
        },
      ],
    }),
  };
};
