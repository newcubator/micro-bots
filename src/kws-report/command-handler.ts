import { ActionType } from "../slack/types/slack-types";
import { getProjects } from "../moco/projects";
import { MocoProject } from "../moco/types/moco-types";

export const commandHandler = async () => {
  const projects = await getProjects({ include_archived: false });
  const options = projects
    .slice(0, 100)
    .filter((project: MocoProject) => project.customer.name.toLowerCase().includes("kws"))
    .map((project: MocoProject) => {
      return {
        value: project.id.toString(),
        text: {
          type: "plain_text",
          text: project.name.substring(0, 75),
          emoji: true,
        },
      };
    });
  console.log(`Loaded ${options.length} projects to select from`);

  if (!options.length) {
    return {
      statusCode: 200,
      body: JSON.stringify({
        response_type: "in_channel",
        text: "Konnte keine Projekte finden.",
      }),
    };
  }

  return {
    statusCode: 200,
    body: JSON.stringify({
      text: "Lock project angefragt",
      blocks: [
        {
          type: "section",
          text: {
            type: "mrkdwn",
            text: "Wähle ein KWS Projekt für den Export aus.",
          },
          accessory: {
            type: "static_select",
            action_id: ActionType.KWS_EXCEL_EXPORT,
            placeholder: {
              type: "plain_text",
              text: "Projekt auswählen...",
              emoji: true,
            },
            options: options,
          },
        },
      ],
    }),
  };
};
