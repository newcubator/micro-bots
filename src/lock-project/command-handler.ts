import { getProjects } from "../moco/projects";
import { ActionType } from "../slack/types/slack-types";

export const commandHandler = async () => {
  const projects = await getProjects({ include_archived: false });
  const options = projects
    .slice(0, 100) // slack allows 100 options max
    .map((project) => {
      return {
        value: project.id.toString(),
        text: {
          type: "plain_text",
          text: project.name,
          emoji: true,
        },
      };
    });
  console.log(`Loaded ${options.length} projects to select from`);

  if (!options.length) {
    return {
      statusCode: 200,
      body: JSON.stringify({
        response_type: "in_channel",
        text: "Konnte keine Projekte finden.",
      }),
    };
  }

  return {
    statusCode: 200,
    body: JSON.stringify({
      text: "Lock project angefragt",
      blocks: [
        {
          type: "section",
          text: {
            type: "mrkdwn",
            text: "Ich sperre gerne f체r dich ein Projekt. W채hle dazu bitte rechts das gew체nschte Projekt aus und ich mache mich sofort an die Arbeit.",
          },
          accessory: {
            type: "static_select",
            action_id: ActionType.LOCK_PROJECT,
            placeholder: {
              type: "plain_text",
              text: "Projekt ausw채hlen...",
              emoji: true,
            },
            options: options,
          },
        },
      ],
    }),
  };
};
