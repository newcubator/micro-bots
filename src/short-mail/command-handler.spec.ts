import { getContacts } from "../moco/contacts";
import { commandHandler } from "./command-handler";

jest.mock("../moco/contacts");
const getContactsMock = getContacts as jest.Mock;

test("command handler for short mail", async () => {
  let result = await commandHandler({} as any);
  expect(result.statusCode).toBe(200);
  expect(JSON.parse(result.body)).toMatchSnapshot();
});

test("respond when loads no contacts", async () => {
  getContactsMock.mockResolvedValueOnce([]);
  let result = await commandHandler({} as any);
  expect(result.statusCode).toBe(200);
  expect(JSON.parse(result.body)).toMatchSnapshot();
});
