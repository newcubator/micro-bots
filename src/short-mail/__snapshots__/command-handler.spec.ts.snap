// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command handler short mail 1`] = `"{\\"response_type\\":\\"in_channel\\",\\"text\\":\\"Kurzbrief angefragt\\",\\"blocks\\":[{\\"type\\":\\"input\\",\\"block_id\\":\\"SHORT_MAIL_RECIPIENT\\",\\"label\\":{\\"type\\":\\"plain_text\\",\\"text\\":\\"Ich erstelle dir gerne einen Kurzbrief. Wähle dazu den Empfänger und deinen Standort aus und gibt deine Nachricht ein.\\"},\\"element\\":{\\"type\\":\\"static_select\\",\\"action_id\\":\\"SHORT_MAIL_RECIPIENT\\",\\"placeholder\\":{\\"type\\":\\"plain_text\\",\\"text\\":\\"Kontakt auswählen...\\",\\"emoji\\":true},\\"options\\":[{\\"value\\":\\"2\\",\\"text\\":{\\"type\\":\\"plain_text\\",\\"text\\":\\"Elon Musk\\",\\"emoji\\":true}},{\\"value\\":\\"2\\",\\"text\\":{\\"type\\":\\"plain_text\\",\\"text\\":\\"Steve Jobs\\",\\"emoji\\":true}}]}},{\\"type\\":\\"input\\",\\"block_id\\":\\"SHORT_MAIL_TEXT\\",\\"element\\":{\\"type\\":\\"plain_text_input\\",\\"action_id\\":\\"SHORT_MAIL_TEXT\\",\\"multiline\\":true},\\"label\\":{\\"type\\":\\"plain_text\\",\\"text\\":\\"Deine Nachricht:\\",\\"emoji\\":true}},{\\"type\\":\\"input\\",\\"block_id\\":\\"SHORT_MAIL_LOCATION\\",\\"label\\":{\\"type\\":\\"plain_text\\",\\"text\\":\\"Wähle deinen Standort aus:\\"},\\"element\\":{\\"type\\":\\"static_select\\",\\"action_id\\":\\"SHORT_MAIL_LOCATION\\",\\"placeholder\\":{\\"type\\":\\"plain_text\\",\\"text\\":\\"Standort auswählen...\\",\\"emoji\\":true},\\"options\\":[{\\"text\\":{\\"type\\":\\"plain_text\\",\\"text\\":\\"Dortmund\\",\\"emoji\\":true},\\"value\\":\\"D\\"},{\\"text\\":{\\"type\\":\\"plain_text\\",\\"text\\":\\"Hannover\\",\\"emoji\\":true},\\"value\\":\\"H\\"}]}},{\\"type\\":\\"actions\\",\\"block_id\\":\\"confirmationButton\\",\\"elements\\":[{\\"type\\":\\"button\\",\\"text\\":{\\"type\\":\\"plain_text\\",\\"text\\":\\"Generieren\\"},\\"value\\":\\"Confirmation\\",\\"style\\":\\"primary\\",\\"action_id\\":\\"SHORT_MAIL\\"}]}]}"`;

exports[`respond when loads no contacts 1`] = `"{\\"response_type\\":\\"in_channel\\",\\"text\\":\\"Konnte keine Kontakte finden.\\"}"`;
